include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,321 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
-  const boost::signals2::signal<void(int)> &reorthogonalize_signal)
+namespace internal
 {
-  Assert(dim > 0, ExcInternalError());
-  const unsigned int inner_iteration = dim - 1;
+  namespace SolverGMRESImplementation
+  {
+    template <class VectorType>
+    void
+    Tvmult_add(const unsigned int dim,
+               const VectorType & vv,
+               const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+                 &             orthogonal_vectors,
+               Vector<double> &h)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] += vv * orthogonal_vectors[i];
+    }
 
-  // need initial norm for detection of re-orthogonalization, see below
-  double     norm_vv_start = 0;
-  const bool consider_reorthogonalize =
-    (reorthogonalize == false) && (inner_iteration % 5 == 4);
-  if (consider_reorthogonalize)
-    norm_vv_start = vv.l2_norm();
 
-  for (unsigned int i = 0; i < dim; ++i)
-    h[i] = 0;
 
-  for (unsigned int c = 0; c < 2; ++c) // 0: orthogonalize, 1: reorthogonalize
+    template <class Number>
+    void
+    Tvmult_add(
+      const unsigned int                                                   dim,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
     {
-      // Orthogonalization
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
+      AlignedVector<VectorizedArray<double>> hs(dim,
+                                                VectorizedArray<double>(0.0));
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
+        for (unsigned int i = 0; i < dim; ++i)
+          DEAL_II_OPENMP_SIMD_PRAGMA
+      for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+        hs[i][v] +=
+          orthogonal_vectors[i].local_element(j + v) * vv.local_element(j + v);
+
+      for (unsigned int i = 0; i < dim; ++i)
+        for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+          h(i) += hs[i][v];
+
+      for (; j < vv.locally_owned_size(); ++j)
+        for (unsigned int i = 0; i < dim; ++i)
+          h(i) += orthogonal_vectors[i].local_element(j) * vv.local_element(j);
+
+      Utilities::MPI::sum(h, MPI_COMM_WORLD, h);
+    }
+
+
+
+    template <class VectorType>
+    double
+    substract_and_norm(
+      const unsigned int dim,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &                   orthogonal_vectors,
+      const Vector<double> &h,
+      VectorType &          vv)
+    {
+      Assert(dim > 0, ExcInternalError());
+
+      for (unsigned int i = 0; i < dim - 1; ++i)
+        vv.add(-h(i), orthogonal_vectors[i]);
```

Comment:
If the other function is called `Tvmult_add`, shouldn't this one be `vmult_add`? And possibly with the norm? Overall, I think we should say that the name `Tvmult` and `vmult` assumes a column-major storage of the `TmpVectors` matrix, right? Which is the BLAS storage order, and probably the best option for a linear algebra interface. My initial thought had been the C++ storage which is row-major, but I agree it is better to choose the BLAS/Fortran order.

Reply:
For me it looks like https://www.dealii.org/developer/doxygen/deal.II/classLinearAlgebra_1_1distributed_1_1Vector.html#a25d752a33b33192932b0dd8054d2c40e (just that arbitrary number of vector is accepted).

Reply:
Sure, but operations that are extended from single vectors (BLAS-1) to act on multiple vectors are essentially BLAS-2 functions. The only real difference is that we do not allocate all memory at once, so we need to toggle an array of pointers than simply accessing a single block of memory with fixed stride. In any case, I would prefer symmetric names between the `vmult` and `Tvmult` operations in namings, i.e., either both or none. I do not have a strong opinion, though, as one is performing the reductions along the long dimension (`dot product`), the other along the short dimensions (`vector additions`), so I can be convinced by a useful name that is not too uncommon in this field.